cmd/goprotoc/codegen.go

```diff
@@ -107,10 +107,27 @@ func doCodeGen(outputs map[string]string, fds []*desc.FileDescriptor, pluginDefs
 		}
 		w, err := os.OpenFile(fileName, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0666)
 		if err != nil {
+			if os.IsNotExist(err) {
+				// The file does not have to exist since we pass o.O_CREATE but the above
+				// open can return error if the directory does not exist, and protoc
+				// makes all necessary subdirectories as part of generation.
+				if err := os.MkdirAll(filepath.Dir(fileName), 0755); err != nil {
```

Comment:
Maybe we should have an initial pass over `outputs` that verifies the indicated directories exist (the root of output). And then down here, we could always call `MkdirAll` _before_ `OpenFile` (so we don't need to try to call `OpenFile` 2x).

Reply:
I think you only want to `MkdirAll` when you actually get to the point of `OpenFile`, not do all `MkdirAlls` and then do the `OpenFiles`. Imagine if you have 100 packages, you create 100 directories, and then the third write to a file fails - now you have 99 empty directories. I'll restructure this to do the `MkdirAll` then a single `OpenFile`.

Reply:
I just meant call `MkdirAll` _right before_ calling `OpenFile`. Not to make all directories in one pass first.

When I meant an initial pass over `outputs`, that would check to see if the directories exist and fail if they do not. (I thought that's what you said that protoc does. Maybe I misunderstood.)

Reply:
Ah yes, this makes sense. Will do.